File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/screen/TrainDetectorScreen.java
package cc.mirukuneko.realtrainmodrenewed.client.screen;

import cc.mirukuneko.realtrainmodrenewed.blockentity.TrainDetectorBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper;
import cc.mirukuneko.realtrainmodrenewed.network.ConfigureTrainDetectorPayload;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;

public class TrainDetectorScreen extends Screen {
    private final BlockPos pos;
    private EditBox channelBox;
    private EditBox rangeBox;
    private int linkedChannel = -1;
    private int detectionRange = 3;
    private boolean occupied;

    public TrainDetectorScreen(BlockPos pos) {
        super(Component.literal("電車検知ブロック"));
        this.pos = pos.immutable();
    }

    @Override
    protected void init() {
        readState();
        int boxWidth = 150;
        int x = (width - boxWidth) / 2;
        int y = height / 2 - 36;
        channelBox = new EditBox(font, x, y, boxWidth, 20, Component.literal("信号番号"));
        channelBox.setMaxLength(10);
        channelBox.setValue(linkedChannel > 0 ? Integer.toString(linkedChannel) : "");
        addRenderableWidget(channelBox);

        rangeBox = new EditBox(font, x, y + 30, boxWidth, 20, Component.literal("検知範囲"));
        rangeBox.setMaxLength(3);
        rangeBox.setValue(Integer.toString(detectionRange));
        addRenderableWidget(rangeBox);
        setInitialFocus(channelBox);

        addRenderableWidget(Button.builder(Component.literal("保存"), button -> submit())
            .bounds(width / 2 - 80, y + 66, 75, 20)
            .build());
        addRenderableWidget(Button.builder(Component.translatable("gui.cancel"), button -> onClose())
            .bounds(width / 2 + 5, y + 66, 75, 20)
            .build());
    }

    private void readState() {
        Minecraft minecraft = Minecraft.getInstance();
        if (minecraft.level == null) {
            return;
        }
        if (minecraft.level.getBlockEntity(pos) instanceof TrainDetectorBlockEntity blockEntity) {
            linkedChannel = blockEntity.getLinkedChannel();
            detectionRange = blockEntity.getDetectionRange();
            occupied = blockEntity.isOccupied();
        }
    }

    private void submit() {
        try {
            int channel = channelBox.getValue().trim().isEmpty() ? -1 : Integer.parseInt(channelBox.getValue().trim());
            int range = Integer.parseInt(rangeBox.getValue().trim());
            ClientNetworkHelper.sendToServer(new ConfigureTrainDetectorPayload(pos, channel, range));
            onClose();
        } catch (NumberFormatException ignored) {
            if (Minecraft.getInstance().player != null) {
                Minecraft.getInstance().player.sendOverlayMessage(Component.literal("数字で入力してください"));
            }
        }
    }

    @Override
    public void extractRenderState(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float partialTick) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick);
        int centerY = height / 2;
        graphics.centeredText(font, title, width / 2, centerY - 62, 0xFFFFFF);
        graphics.centeredText(font, Component.literal("占有中: " + (occupied ? "はい" : "いいえ")), width / 2, centerY - 50, occupied ? 0xFF6666 : 0x66FF66);
        graphics.centeredText(font, Component.literal("信号番号を入れると occupied=停止 / clear=進行"), width / 2, centerY - 18, 0xAAAAAA);
        graphics.centeredText(font, Component.literal("空欄なら赤石検知だけ使えます"), width / 2, centerY - 6, 0xAAAAAA);
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/screen/TrainDetectorScreen.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client.screen

import cc.mirukuneko.realtrainmodrenewed.blockentity.TrainDetectorBlockEntity
import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper
import cc.mirukuneko.realtrainmodrenewed.network.ConfigureTrainDetectorPayload
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.gui.screens.Screen
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component

open class TrainDetectorScreen(pos: BlockPos) : Screen(Component.translatable("block.realtrainmodrenewed.train_detector")) {
    private val pos: BlockPos = pos.immutable()
    private lateinit var channelBox: EditBox
    private lateinit var rangeBox: EditBox
    private var linkedChannel: Int = -1
    private var detectionRange: Int = 3
    private var occupied: Boolean = false

    override fun init() {
        readState()
        val boxWidth = 150
        val x = (width - boxWidth) / 2
        val y = height / 2 - 36
        channelBox = EditBox(font, x, y, boxWidth, 20, Component.translatable("screen.realtrainmodrenewed.signal.channel"))
        channelBox.setMaxLength(10)
        channelBox.setValue(if (linkedChannel > 0) linkedChannel.toString() else "")
        addRenderableWidget(channelBox)

        rangeBox = EditBox(font, x, y + 30, boxWidth, 20, Component.translatable("screen.realtrainmodrenewed.train_detector.range"))
        rangeBox.setMaxLength(3)
        rangeBox.setValue(detectionRange.toString())
        addRenderableWidget(rangeBox)
        setInitialFocus(channelBox)

        addRenderableWidget(
            Button.builder(Component.translatable("button.realtrainmodrenewed.save")) { submit() }
                .bounds(width / 2 - 80, y + 66, 75, 20)
                .build()
        )
        addRenderableWidget(
            Button.builder(Component.translatable("gui.cancel")) { onClose() }
                .bounds(width / 2 + 5, y + 66, 75, 20)
                .build()
        )
    }

    private fun readState() {
        val minecraft = Minecraft.getInstance()
        if (minecraft.level == null) {
            return
        }
        val blockEntity = minecraft.level!!.getBlockEntity(pos)
        if (blockEntity is TrainDetectorBlockEntity) {
            linkedChannel = blockEntity.linkedChannel
            detectionRange = blockEntity.detectionRange
            occupied = blockEntity.isOccupied()
        }
    }

    private fun submit() {
        try {
            val channel = if (channelBox.value.trim().isEmpty()) -1
                else channelBox.value.trim().toInt()
            val range = rangeBox.value.trim().toInt()
            ClientNetworkHelper.sendToServer(ConfigureTrainDetectorPayload(pos, channel, range))
            onClose()
        } catch (ignored: NumberFormatException) {
            if (Minecraft.getInstance().player != null) {
                Minecraft.getInstance().player!!.sendOverlayMessage(
                    Component.translatable("message.realtrainmodrenewed.number_required")
                )
            }
        }
    }

    override fun extractRenderState(
        graphics: GuiGraphicsExtractor,
        mouseX: Int,
        mouseY: Int,
        partialTick: Float
    ) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick)
        val centerY = height / 2
        graphics.centeredText(font, title, width / 2, centerY - 62, 0xFFFFFF)
        graphics.centeredText(
            font,
            Component.translatable(
                "screen.realtrainmodrenewed.train_detector.occupied",
                Component.translatable(if (occupied) "gui.yes" else "gui.no"),
            ),
            width / 2,
            centerY - 50,
            if (occupied) 0xFF6666 else 0x66FF66
        )
        graphics.centeredText(
            font,
            Component.translatable("screen.realtrainmodrenewed.train_detector.signal_hint"),
            width / 2,
            centerY - 18,
            0xAAAAAA
        )
        graphics.centeredText(
            font,
            Component.translatable("screen.realtrainmodrenewed.train_detector.redstone_hint"),
            width / 2,
            centerY - 6,
            0xAAAAAA
        )
    }

    override fun isPauseScreen(): Boolean = false
}
